### PR TITLE
Fix: processing react context listeners after the JS loads on Android

### PR DESCRIPTION
**Pull request by coado**
## Summary:

According to this [issue](https://github.com/facebook/react-native/issues/46306) the event is emitted before the listener is attached on JS side when the push notification is clicked. The user sends an event when the `ReactContext` is initialized by adding `onReactContextInitialized` listeners on `ReactInstanceManager`, which invokes `sendEvent` function. This mechanism worked on the old architecture, but after migrating to Fabric it stopped. There are two main differences between the old and the new architecture here:

1. On the new architecture the `onReactContextInitialized` listeners should be added to `ReactHost` instead,
2. `onReactContextInitialized` listeners are invoked before the JS loads

We can mitigate this issue by running a new task after the JS starts preparing itself, move processing `onReactContextInitialized` listeners there, and run them on the JS thread.

My questions are if its okay to call all such listeners on the JS thread and is it okay to run them in this particular moment?

## Changelog:

[ANDROID] [FIXED] - Move processing onReactContextInitialized listeners after the JS starts loading and run them on the JS thread.

## Test Plan:

Tested on the repro https://github.com/coado/react-native-event-emitter-issue and the rn-tester by registering listeners on the `ReactHost`.

### CODE REVIEW

**Comment by javache** [comment]
On the old architecture, we call this event emitter from the UI thread, we should maintain this behaviour.

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/runtime/ReactHostImpl.java [hunk 1]
```diff
@@ -1000,6 +1000,23 @@ private Task<Void> callAfterGetOrCreateReactInstance(
               return null;
             },
             executor)
+        .continueWithTask(
+          task -> {
+            final BridgelessReactContext reactContext = getOrCreateReactContext();
+
+            reactContext.runOnJSQueueThread(
```
Comment (javache) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/runtime/ReactHostImpl.java:1007]: This causes this event to escape the async Task setup workflow, and will lead to inconsistencies if we're also trying to teardown.

Additionally, you've added this to `callAfterGetOrCreateReactInstance`, which means the event will be called many multiple times during the lifecycle which is incorrect.

**Review by javache — CHANGES_REQUESTED** [review]

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/runtime/ReactHostImpl.java (hunk 1 above)
Comment (coado) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/runtime/ReactHostImpl.java:1007]: Okay that's absolutely terrible 😅 I will think about doing it the other way!

**Comment by coado** [comment]
Doing something like this still doesn't fully solve the issue because of the potential race condition caused by the concurrent root on fabric. The root may be initialized after the event is emitted (listeners are attached in the main component, check [repro](https://github.com/coado/react-native-event-emitter-issue)). If the `useConcurrentRootOverride` is set to `false` in `renderApplication.js` it should be working. I am not sure if there is a way to run native to JS calls after the root is completed and if we even want to call `onReactContextInitialized` listeners after that 🤔

**Comment by javache** [comment]
This is not the right direct. `onReactContextInitialized` should be co-located with any surface lifecycle management.

`ReactInstanceEventListener`s are invoked when it is safe to execute new JS calls, can you clarify what happens when you say "onReactContextInitialized listeners are invoked before the JS loads"? Are you seeing events emitted never received?

**Comment by coado** [comment]
> This is not the right direct. `onReactContextInitialized` should be co-located with any surface lifecycle management.
> 
> `ReactInstanceEventListener`s are invoked when it is safe to execute new JS calls, can you clarify what happens when you say "onReactContextInitialized listeners are invoked before the JS loads"? Are you seeing events emitted never received?

I meant that `onReactContextInitialized` event listeners are [invoked](https://github.com/facebook/react-native/blob/main/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/runtime/ReactHostImpl.java#L1186-L1190) before the surface [starts](https://github.com/facebook/react-native/blob/main/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/runtime/ReactHostImpl.java#L239) which is opposite to what happened on the old architecture and it causes the above bug (emitted events are never received).

**Comment by javache** [comment]
I think this is working as intended then. If you have listeners that you want to to setup immediately after the JS has finished loading, you should set them up outside of the context of a React root as part of your bundle entrypoint.

If anything, we could move surface start inside a `onReactContextInitialized` callback - react context initialization doesn't require surfaces to have started.

**Pull request closed without merging**